### PR TITLE
PP-1059: Jobs in queues with resources_available limits (e.g., resvs)…

### PR DESCRIPTION
**Pull request by bhroam**
… may not run due to equiv classes

#### Issue-ID

* **[PP-1059](https://pbspro.atlassian.net/browse/PP-1059)**

#### Problem description
* When a queue has a resources_available limit, it can cause other similar jobs to not run.

#### Cause / Analysis
* When a job hits a queue's resources_available limit, the equivalence class it is in is marked can not run.  Jobs in other queues can be in that class.  If they are, they'll also not run even though their queue does not have a resources_available limit.  This is because the class is already marked can not run.

#### Solution description
* Jobs in different queues can be in the same equivalence class when the queue does not matter.  Jobs in queues with resources_available or jobs in reservation queues need to be in their own equivalence class.

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/scheduler/data_types.h [hunk 1]
```diff
@@ -372,6 +372,7 @@ struct queue_info
 	unsigned has_soft_limit:1;	/* queue has a soft user/grp limit set */
 	unsigned has_hard_limit:1;	/* queue has a hard user/grp limit set */
 	unsigned is_peer_queue:1;	/* queue is a peer queue */
+	unsigned has_res_av:1;		/* queue has resources_available limits */
```
Comment (agrawalravi90) [inline src/scheduler/data_types.h]: I think we should add the word 'limit' to the identifier, maybe 'has_resav_limit' ?

File: test/tests/functional/pbs_equiv_classes.py [hunk 2]
```diff
@@ -1840,3 +1840,44 @@ def test_held_jobs_equiv_class(self):
...
+        a = {'queue': 'workq', 'Resource_List.select': '1:ncpus=1'}
+        J = Job(TEST_USER, a)
+        jid1 = self.server.submit(J)
+
+        a['queue'] = 'workq2'
+        J = Job(TEST_USER, a)
+        jid2 = self.server.submit(J)
+
+        J = Job(TEST_USER)
+        jid3 = self.server.submit(J)
```
Comment (agrawalravi90) [inline test/tests/functional/pbs_equiv_classes.py]: I don't like relying on the fact that PBS/PTL will add a default select to the job and would suggest that you add the select line explicitly.

File: src/scheduler/data_types.h (hunk 1 above)
Comment (minghui-liu) [reply src/scheduler/data_types.h]: +1 on changing the name to `has_res_av_limit`.

Comment (bhroam) [reply src/scheduler/data_types.h]: I changed the name.

File: test/tests/functional/pbs_equiv_classes.py (hunk 2 above)
Comment (bhroam) [reply test/tests/functional/pbs_equiv_classes.py]: I meant to include it.  I just missed it, thanks for catching this.

**Comment by bhroam** [comment]
I modified the test a little.  I'm not sure why it was passing before.  Job 2 runs and Job 3 doesn't.  I was checking for them in the other order.  I also removed the expect() for the comment because it can change due to a race condition.  The initial cycle it gets set to the comment about not enough queue resources.  Subsequent cycles the node is full so it changes to not enough free nodes.

**Comment by agrawalravi90** [comment]
Thanks for addressing the comments, could you please attach the latest testing logs to the ticket?

**Comment by bhroam** [comment]
@agrawalravi90 I attached the test logs and squashed.  When running the entire equivalence class test suite, I found another bug.  I filed it as PP-1127.

**Comment by agrawalravi90** [comment]
Thanks Bhroam, I'll merge this in

**Pull request merged**